### PR TITLE
[ATL] Make CImage thread-safe

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Improve compatibility.
JIRA issue: [CORE-19008](https://jira.reactos.org/browse/CORE-19008)

## Proposed changes

- Add `CInitGDIPlus` subclass in `CImage`.
- Lock the `CInitGDIPlus` data by using a critical section.
- Add `CImage::InitGDIPlus` and use it.

## TODO

- [x] Do small tests.

## Screenshots

BEFORE:
before

AFTER:
after

after-2

### CODE REVIEW

File: sdk/lib/atl/atlimage.h [hunk 1]
```diff
@@ -58,25 +55,26 @@ class CImage
...
 
-    operator HBITMAP()
+    operator HBITMAP() throw()
     {
         return m_hbm;
     }
 
-public:
-    void Attach(HBITMAP hBitmap, DIBOrientation eOrientation = DIBOR_DEFAULT)
+    static void ReleaseGDIPlus()
```
Comment (HBelusca) [inline sdk/lib/atl/atlimage.h:72]: That one should be public?

Comment (katahiromz) [reply sdk/lib/atl/atlimage.h:72]: There's `public:` above here.

File: sdk/lib/atl/atlimage.h [hunk 2]
```diff
@@ -846,153 +848,186 @@ class CImage
...
-                hinstGdiPlus = NULL;
+                ReleaseGDIPlus();
             }
+            ::LeaveCriticalSection(&m_sect);
         }
-    }; // struct COMMON
+    };
 
-    static COMMON*& GetCommonPtr()
+    static CInitGDIPlus* GetInitGDIPlusInstance()
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply sdk/lib/atl/atlimage.h:1014]: This is compatible with the original header.

[the remark this answers is not in the document]
Comment (HBelusca) [reply sdk/lib/atl/atlimage.h:1014]: Mine doesn't contain it.

[the remark this answers is not in the document]
Comment (katahiromz) [reply sdk/lib/atl/atlimage.h]: Simplified.

[the remark this answers is not in the document]
Comment (katahiromz) [reply sdk/lib/atl/atlimage.h]: `inline static` variables are not available before C++17.

[the remark this answers is not in the document]
Comment (HBelusca) [reply sdk/lib/atl/atlimage.h]: Nobody mentioned anything about "inline static". I was just mentioning the usual static members. (The MS ATL header that directly use those is the ATL from MSVC 2010, which is of course not aware of C++11, and obviously  not C++17.)

[the remark this answers is not in the document]
Comment (katahiromz) [reply sdk/lib/atl/atlimage.h]: The non-inline static variable in class needs its definition. That definition cannot be done in header only code.

[the remark this answers is not in the document]
Comment (katahiromz) [reply sdk/lib/atl/atlimage.h]: For example, the following code:

```
struct A
{
static int s_i;
};
```
needs `A::s_i` definition.

[the remark this answers is not in the document]
Comment (katahiromz) [reply sdk/lib/atl/atlimage.h]: image

[the remark this answers is not in the document]
Comment (HBelusca) [reply sdk/lib/atl/atlimage.h]: Ah yes true, and that's indeed what they did (where an atlimage2.cpp is used for that).
But for information I've also looked at other such cpp files present in the MS ATL to see for what they were used, and it turned out it is also for such static variables to be defined.
Then I looked at what we did, and saw that we do this instead:
https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&s=%28selectany%29
(see the results in the ATL).
We have those variables defined, but in the header and with that declspec.

[the remark this answers is not in the document]
Comment (katahiromz) [reply sdk/lib/atl/atlimage.h]: Used `DECLSPEC_SELECTANY`. Thanks.

**Comment by katahiromz** [comment]
Please wait patiently for one more day.

**Pull request merged**